plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp

```diff
@@ -0,0 +1,59 @@
+#pragma once
+#include <stdint.h>
+#include <string_view>
+#include <eosio/chain_plugin/chain_plugin.hpp>
+#include <eosio/blockvault_client_plugin/blockvault.hpp>
+
+namespace eosio {
+    namespace blockvault {
+
+        template<typename BP>
+        struct blockvault_sync_strategy : public sync_callback {
+            blockvault_sync_strategy(block_vault_interface* blockvault, BP& blockchain_provider, std::function<void()> shutdown,
+                                     std::function<bool()> check_shutdown) :
+                _previous_block_id_specified(false), _blockvault(blockvault), _blockchain_provider(blockchain_provider), _shutdown(shutdown), _check_shutdown(check_shutdown) {
+
+            }
+
+            blockvault_sync_strategy(block_vault_interface* blockvault, BP& blockchain_provider, std::function<void()> shutdown,
+                                     std::function<bool()> check_shutdown, chain::block_id_type block_id) :
+                _previous_block_id_specified(true), _blockvault(blockvault), _blockchain_provider(blockchain_provider), _shutdown(shutdown), _check_shutdown(check_shutdown),
+                _previous_block_id(block_id) {
+            }
+
+            void do_sync() {
+                _blockvault->sync(_previous_block_id_specified ?
+                    &_previous_block_id : nullptr, *this);
+            }
+
+            void on_snapshot(const char* snapshot_filename) override final {
+                auto infile = std::ifstream(snapshot_filename, (std::ios::in | std::ios::binary));
+                auto reader = std::make_shared<chain::istream_snapshot_reader>(infile);
+                _blockchain_provider.chain->startup(_shutdown, _check_shutdown, reader);
+                infile.close();
+
+                _got_snapshot = true;
+            }
+
+            void on_block(eosio::chain::signed_block_ptr block) override final {
+                if (!_got_snapshot) {
```

Comment:
Undefined behavior if this is called before `on_snapshot`. `_got_snapshot` is not initialized.